Proposed refactor / changes

ibhagwan opened the pull request:
Hi @elanmed,

First, ty for writing this wonderful project, I didn't think fzf-lua was missing anything but this is quite cool and well done and I'm going to add this to my personal arsenal :-)

I see this as a  potential replacement for `files|oldfiles` and wanted to discuss with you the proposed changes.

This is still WIP (testing, etc) as I wanted to first get your feedback before I remove more parts of the code and make further adjustments.

The biggest change would be dropping the `cwds` folder and separate databases for each `cwd`, this makes us cwd dependent and too limiting, if I change cwd to a project's sub directory or wish to see a full picture of all recent files (a-la `oldfiles`).

With the proposed changes `fn_transform` does the filtering based on the current `cwd` so we can handle the combined sorted files database and `cwd_only` (which is already an option in fzf-lua used with any file-type picker) can control the display.

I currently piggybacked the `cwd_only` option to decide if we show the scoreless files after the sorted files but this probably needs a separate option.

My vision is this:
1. Replacement for `files` combined with `oldfiles cwd_only=true`
```lua
:lua require("fzf-lua-frecency").frecency({cwd_only=true})
```
<img width="1019" height="571" alt="image" src="https://github.com/user-attachments/assets/c9a568d8-f4f5-4d23-8dce-8cc94724338e" />


2. Replacement for the global `oldfiles` which shows all recent files without appending the `fd` command:
```lua
:lua require("fzf-lua-frecency").frecency({})
```

<img width="1028" height="546" alt="image" src="https://github.com/user-attachments/assets/b63cdf07-94a1-4de5-86a8-0b4c3f0a57a1" />


In addition we no longer use the `wrapped_enter` as I find this also limits updating the scores only when using fzf-lua-frecency, instead the user should call setup once to setup the BufWinEnter autocmd (we can also call the setup automatically for the user once on the initial load, up to discussion).

Overall, there are a few more changes I'd like to add in fzf-lua as well, a proper interface for registering the extension which will add `:FzfLua frecency` command so users don't need to call `require("fzf-lua-frecency")`.

Let me know your thoughts?

elanmed commented:
> First, ty for writing this wonderful project, I didn't think fzf-lua was missing anything but this is quite cool and well done and I'm going to add this to my personal arsenal :-)

Thank you! :smile: 

> The biggest change would be dropping the cwds folder and separate databases for each cwd, this makes us cwd dependent and too limiting, if I change cwd to a project's sub directory or wish to see a full picture of all recent files (a-la oldfiles).

I like this, I didn't think about viewing all the frecent files from all directories at once, that's a cool idea

> :lua require("fzf-lua-frecency").frecency({cwd_only=true})

I wonder if another option like `all_files` could be helpful - something like:
```lua
require("fzf-lua-frecency").frecency({cwd_only=true}) -- oldfiles in the current dir
require("fzf-lua-frecency").frecency({}) -- oldfiles in all dirs
require("fzf-lua-frecency").frecency({cwd_only=true, all_files=true}) -- oldfiles + all files in the current dir
require("fzf-lua-frecency").frecency({all_files}) -- oldfiles + all files in all dirs
```

I could see `oldfiles in the current dir` being useful :thinking: 

> In addition we no longer use the wrapped_enter as I find this also limits updating the scores only when using fzf-lua-frecency, instead the user should call setup once to setup the BufWinEnter autocmd (we can also call the setup automatically for the user once on the initial load, up to discussion).

This makes sense to me! One other thing I do in `wrapped_enter` is remove the display score before passing that onto `fzf_lua.path.entry_to_file`. 

https://github.com/elanmed/fzf-lua-frecency.nvim/blob/22b159ecb268bc4e2b1d69bc2eebc85a20a88b33/lua/fzf-lua-frecency/init.lua#L42-L44

I noticed that `fzf_lua.path.entry_to_file` wouldn't remove the display score if there were also no file icons - but I can't reproduce that anymore, `fzf_lua.path.entry_to_file` might have been updated or I might have hallucinated.

> Btw, I also fixed multiprocess which wasn’t working also made sure the transform function is generated once and not with every file (get_fn_transform was called for every entry).

Hah I'm curious what I did wrong with `multiprocess`. Good to know that `fn_transform` can take a string, thanks for catching! :pray:

ibhagwan commented:
> One nice thing about storing the scores per cwd is that I can also calculate the max score per cwd, which I use for padding the entries so all the files align. It would probably be simplest to add a fixed padding, something like 5 digits 🤔

If that’s the only reasoning 5-6 digits should be more than enough, if the score is high (over a 100) we can clamp the fractional digits instead?

> I wonder if another option like all_files could be helpful - something like:

I like the `all_files` option, flexible enough for all combos of dir/files.


> I noticed that fzf_lua.path.entry_to_file wouldn't remove the display score if there were also no file icons - but I can't reproduce that anymore, fzf_lua.path.entry_to_file might have been updated or I might have hallucinated.

You’re not hallucinating, entry_to_file will strip the string up to `FzfLua.utils.nbsp`, whereas you used a simple space, when icons are present the nbsp is placed post icon so the score is stripped as part of that, if you place one nbsp at the end or the padding after the score string it will work with or without icons/formatting/anything?

> Hah I'm curious what I did wrong with multiprocess. Good to know that fn_transform can take a string, thanks for catching! 🙏

There was nothing wrong per-se, this “know how” is specific to fzf-lua, the code was using `require` for both fzf-lua and fzf-lua-frecency, that doesn’t work in the headless instance as there are no plugins loaded, as you see from the code I added runtime path string so that requiring fzf-lua-frecency won’t fail, I also replaced require fzf-lua with the global `FzfLua` which is created and maintained by the headless instance.

By modifying the fn_transform to return an “parameterized” closure (with the rpc_opts) we instantiate the function only once instead of with every call, could work with string or without it was just simpler this way because you already used a string function due to mpack.

Let me know how you’d like me to proceed with this PR, if you are ok with one DB for everything I can remove the cwd indexing of max_score and dated files.

elanmed commented:
> If that’s the only reasoning 5-6 digits should be more than enough, if the score is high (over a 100) we can clamp the fractional digits instead?

Clamping the fractional digits is a cool idea too

> the code was using require for both fzf-lua and fzf-lua-frecency, that doesn’t work in the headless instance as there are no plugins loaded

Ah interesting, makes sense

> By modifying the fn_transform to return an “parameterized” closure (with the rpc_opts) we instantiate the function only once instead of with every call,

Oh that's really cool, interesting!

> Let me know how you’d like me to proceed with this PR, if you are ok with one DB for everything I can remove the cwd indexing of max_score and dated files.

I'm good with one DB! Feel free to go ahead with the PR :pray:

ibhagwan commented:
> I'm good with one DB! Feel free to go ahead with the PR 🙏

Great, I’ll work on it :)

A few other considerations:

- Do we want to support windows? If so we need to replace the awk with a dedup table internally inside `fn_transform` (we can store it in the opts), or maybe require windows to install awk?
- Currently fzf-lua can work with fd/rg/find - if we don’t want to support all we have to test for fd executable and alert the user if it needs installing.

ibhagwan commented:
Still didn't refactor the single db but it's working since I hardcoded `cwd="/"`, I will do that the next bunch of changes.

I did refactor this heavily after adding https://github.com/ibhagwan/fzf-lua/commit/7ada0850880ddd17709c90990697e30e0b27008c in fzf-lua which enables proper extension registration.

Both the picker and options are now registered to fzf-lua so you get the benfit of the `FzfLua` command and auto-complete, I also moved the `fzf_lua_frecency` opts into the fzf-lua opts so it's easier to use (you can run `:FzfLua frecency display_score=false`, etc):
> completion of `FzfLua ...`
<img width="683" height="243" alt="image" src="https://github.com/user-attachments/assets/a805d4ba-a748-4314-b689-4c8074916bb4" />

> completion of `FzfLua frecency ...` shows fzf-lua option and fzf-lua-frecency combined
<img width="722" height="335" alt="image" src="https://github.com/user-attachments/assets/e0e87c9d-6e34-4ad6-a5c9-e335f116c610" />


I also removed `basedir` from `fd` as it's not required (IMHO) and added a fallback for `rg`, while we're at it why not make it a full replacement for files with the option to toggle hidden/follow/no-ignore? checkout the title flags and experiment with toggling using `alt-i|f|h`:
> `:FzfLua frecency cwd_only=true hidden=true no_ignore=true follow=true debug=1`
<img width="1030" height="290" alt="image" src="https://github.com/user-attachments/assets/28722913-26a9-4f75-a3c9-e017f3be3c8b" />

After pressing `alt-f+alt-i`:
<img width="1028" height="285" alt="image" src="https://github.com/user-attachments/assets/c044f5bc-c52d-4fa7-8361-b0b2de9ef97f" />

ibhagwan commented:
Btw, unless specifically set to true or false, the `all_files` option is set automatically to `true` when `cwd_only` is true (still possible ti manually override):
```lua
:FzfLua frecency -- all_files=false
:FzfLua frecency  cwd_only=true -- all_files=true
:FzfLua frecency  cwd_only=true all_files=false
```

I also changed `display_score` default to true as I find it is valuable, similar to `zoxide` interactive in the shell or `:FzfLua zoxide`

Let me know if this is acceptable.

ibhagwan commented on lua/fzf-lua-frecency/fn_transform.lua:
```diff
@@ -42,7 +40,7 @@ M.get_fn_transform = function(opts)
         formatted_score = h.pad_str(h.exact_decimals(score, 2), max_score_len)
       end
-      return ("%s %s"):format(formatted_score, entry)
+      return ("%s%s%s"):format(formatted_score, FzfLua.utils.nbsp, entry)
```
Using `FzfLua.utils.nbsp` enables us to skip stripping the score, `FzfLua.path.entry_to_file` will take care of that automatically.

ibhagwan commented:
One more note, we should probably add something similar to `file_ignore_patterns` so we can exclude certain files like `.git/COMMIT_EDITMSG`, etc.

elanmed commented on lua/fzf-lua-frecency/init.lua:
```diff
@@ -10,136 +17,160 @@ local M = {}
+      -- fzf-lua-frecency defaults
+      display_score = true,
+      cwd_only = false,
+      all_files = function(o) return o and o.cwd_only end,
```
What's your thinking in making `all_files` a function vs a bool @ibhagwan

ibhagwan replied:
So that it can be automatically true if not specified and cwd_only=true, now that I think of it i can remove this function and add `opts.all_files = opts.all_files == nil and cwd_only or opts.all_files`.

I’ll remove it next commit.

elanmed commented:
> Do we want to support windows?
> Currently fzf-lua can work with fd/rg/find - if we don’t want to support all we have to test for fd executable and alert the user if it needs installing.

Good points, created https://github.com/elanmed/fzf-lua-frecency.nvim/issues/2

> I did refactor this heavily after adding https://github.com/ibhagwan/fzf-lua/commit/7ada0850880ddd17709c90990697e30e0b27008c in fzf-lua which enables proper extension registration.'

Very cool! I'm not too familiar with the `_*` opts so I'm glad that you know what's necessary there

> I also moved the fzf_lua_frecency opts into the fzf-lua opts

Sounds good to me, I was afraid of potentially clashing with fzf-lua, but I agree that it's nice to pass options directly to `:FzfLua frecency`

> while we're at it why not make it a full replacement for files with the option to toggle hidden/follow/no-ignore?

That's a cool idea, I like that

> Btw, unless specifically set to true or false, the all_files option is set automatically to true when cwd_only is true (still possible ti manually override):

Makes sense to me :pray:

> I also changed display_score default to true as I find it is valuable, similar to zoxide interactive in the shell or :FzfLua zoxide

This sounds good to me, I think I originally defaulted to `false` to avoid the all the IO overhead back when `multiprocess` was `false`

ibhagwan commented:
Ty for the feedback @elanmed!

I’ll polish it on the next batch of changes and fix the testing too, it’s not far :)

It’s really cool, I feel this can replace both files and oldfiles together, the side effects of using make_entry.file is that this works out of the box with cwd,cwd_only,formatter,path_shorten,file_ignore_patterns and others I’m probably forgetting rn.

One more thing I’d like to consult with you is about expiring non-existing files, we do have the ctrl-x to remove files, we could also add `stat_file` similar to what we have with oldfiles which tests the file’s existence with vim.uv.fs_stat, the downside is it has a somewhat significant performance impact.

ibhagwan commented:
> These performance questions are a little hard for me to gauge - there's some IO when calling frecency to read the DB and generate the latest scores, but is that "alright" because of multiprocess=true? Would checking vim.uv.fs_stat also be alright because of multiprocess=true?

With `multiprocess=true` even if the check is somewhat slow it will never hang or slow down the UI so the UX is always wonderful, it might be slower to add entries when there are 10000 or something but I guess that's ok.

In the last commit I added `stat_file=true` as a default which will test the file using `vim.uv.fs_stat` (also unsure if faster than `vim.fn.filereadable`, no biggie either way), if a file no longer exists, it will also be removed from the database by scheduling `algo.update_file_score`.

Also updated the README (could probably use a bit more polish) and fixed the tests, lint passes as well.

Overall I'm very happy with the result for my own use, I have two mappings:
1. `<leader>fH` -> `FzfLua frecency cwd_only=false all_files=false` to display all "oldfiles" with score without enumerating the cwd
2. `<leader>fh` -> `FzfLua frecency cwd_only=true all_files=true` to display "oldfiles" only in current working directory + enumerate the rest of the files

I'll test replacing my main `<ctrl-p>` mapping for `:FzfLua files`, will have to try with and without `--no-sort` and see which works better as `Files` -  i.e.  `:FzfLua frecency cwd_only=true fzf_opts.--no-sort=false`.

A few notes:
1. Setup is called automatically with no opts on the first picker run, might be worth mentioning it installed an autocmd and that options can be set globally via `setup({ display_score = false})`, etc.
2. No support for Windows ATM, requires minimal changes, might be worth it one day, not in my priority list ATM...
3. We can add a fallback for the default `find` command, also no biggie because who doesn't have at least rg or fd installed nowadays
4. Skim is supported, although we use `--scheme=path`, `normalize_opts` will remove the option for skim and older fzf versions before the option existed, so this works out of the box `:FzfLua frecency profile=skim`.


Area of improvement:
`fn_transform` reads the dated_files/max_scores files **for every entry**, this is a perfect use case for `fn_preprocess`, if I have time tomorrow I'll add it unless you wanna tackle it yourself.

ibhagwan commented:
> fn_transform reads the dated_files/max_scores files for every entry, this is a perfect use case for fn_preprocess, if I have time tomorrow I'll add it unless you wanna tackle it yourself.

Fixed: https://github.com/elanmed/fzf-lua-frecency.nvim/pull/1/commits/6ed030d114819eab7f8dfad7fc5ab43c2afa12f5

ibhagwan commented:
FYI, to test after the latest changes delete `~/.local/share/nvim/fzf-lua-frecency`  or the old db files will interfere.

elanmed commented on lua/fzf-lua-frecency/algo.lua:
```diff
@@ -63,25 +62,24 @@ M.update_file_score = function(filename, opts)
   local fs = require "fzf-lua-frecency.fs"
   local h = require "fzf-lua-frecency.helpers"
-  local cwd = h.default(opts.cwd, vim.fn.getcwd())
+  local db_index = 1 -- We only use index 1 everywhere
```
Do we need `db_index` at all? I think we can remove it and use `dated_files[filename]` directly

elanmed commented on lua/fzf-lua-frecency/init.lua:
```diff
@@ -1,145 +1,193 @@
+  require "fzf-lua"
+
+  -- Register as an fzf-lua extenstion, merge default opts with users' setup opts
+  FzfLua.register_extension("frecency", M.frecency, vim.tbl_deep_extend("keep", opts or {}, {
```
Do you think `force` would be better than `keep` to prefer the user's opts vs the default?

ibhagwan replied:
> Do you think force would be better than keep to prefer the user's opts vs the default? (This assumes the opts from setup are passed along to frecency)

Setup options would lose their relevance in such case, imagine `setup({ display_score=false })` hiding the score globally, setting this to `force` would override the users' wishes and restore the defaults.

elanmed replied:
Ah my bad I misread this code, looks good!

elanmed commented:
> Also updated the README (could probably use a bit more polish)

README looks good for now! I want to polish it a bit too with some info on using the `algo` module directly

> I'll test replacing my main <ctrl-p> mapping for :FzfLua files, will have to try with and without --no-sort and see which works better as Files - i.e. :FzfLua frecency cwd_only=true fzf_opts.--no-sort=false.

I noticed that I actually prefer `--no-sort=false` also :thinking: 

> Setup is called automatically with no opts on the first picker run, might be worth mentioning it installed an autocmd and that options can be set globally via setup({ display_score = false}), etc.

I left a comment about this too, where it might be easier to talk about this, but I'm curious why not pass the `opts` from `frecency` along to `setup`?

> Skim is supported

Ah interesting, I haven't used skim so I wasn't really thinking of this, thanks for taking care of it :pray::pray:

ibhagwan commented on lua/fzf-lua-frecency/algo.lua:
```diff
@@ -132,14 +130,14 @@ M.update_file_score = function(filename, opts)
   }
   local max_scores = fs.read(max_scores_path)
-  max_scores[cwd] = max_score
+  max_scores[db_index] = max_score
```
Same as the above comment, left for posterity, I can leave it or remove it, up to you.

ibhagwan commented on lua/fzf-lua-frecency/fn_transform.lua:
```diff
@@ -1,35 +1,56 @@
 
-    local dated_files_path = h.get_dated_files_path(opts.db_dir)
-    local max_scores_path = h.get_max_scores_path(opts.db_dir)
+    if not _G._fzf_lua_frecency_dated_files then
```
At first I put these in `fn_preprocess` but I found by doing this in `fn_transform` I can get away with a shorter multiprocess command due to a shorter encoded base64, doesn't matter much on Linux/Mac but might matter on Windows where there's a limit to command length due to `cmd.exe`.

ibhagwan commented on lua/fzf-lua-frecency/init.lua:
```diff
@@ -1,145 +1,193 @@
+M.frecency = function(opts)
+  -- Does nothing if already called, will lazy load fzf-lua
+  -- and create the FzfLua global object
+  M.setup()
```
> Should the opts passed to frecency also be passed to setup?

I thought about it and opted against it, assuming the user never called `setup` and ran `:FzfLua frecency display_score=false`, that would override the default value and if the user then runs `:FzfLua frecency` the score will be hidden which may seem confusing due to the implied new default hidden from the user.

ibhagwan commented on lua/fzf-lua-frecency/fn_transform.lua:
```diff
@@ -1,35 +1,56 @@
+    local date_at_score_one = dated_files[db_index] and dated_files[db_index][abs_file] or nil
+
+    -- Only "stat" files from the db, fd|rg enumerated files guaranteed to exist
+    if rpc_opts.stat_file and date_at_score_one then
```
Changed to `vim.uv.fs_stat`.

elanmed replied:
> We could add that but the way I see it update_score is only called from already guaranteed to be existing files?

That's true, but since all the files are sorted during `update_score` we could end up sorting another file (a file that `update_score` wasn't passed) that doesn't exist.

FWIW I also like filtering in `update_score` along with `frecency` in case someone wants to use the `algo` portions of the plugin without the fzf lua portions. 

Current changes look good!

elanmed commented:
PR looks good! I'll merge later tonight when I have some time, thanks for the contributions! :pray::smile:

ibhagwan commented:
https://github.com/elanmed/fzf-lua-frecency.nvim/pull/1/commits/b6d909b1bcc7a3ea363cba0a3539a30c7e10f185 - squashed.
